File: elixir/connect/connect.exs
defmodule Connect do

  @doc """
  Calculates the winner (if any) of a board
  using "O" as the white player
  and "X" as the black player
  """
  @spec result_for([String.t]) :: :none | :black | :white
  def result_for(board) do
    goal_x = (board |> Enum.at(0) |> String.length) - 1
    cond do
      board |> pieces_for("X") |> winner?(goal_x) -> :black
      board |> pieces_for("O") |> winner?(goal_x) -> :white
      true -> :none
    end
  end

  defp pieces_for(board, player) do
    board
    |> transpose(player)
    |> build_board
    |> Enum.filter(fn {_, _, v} -> v == player end)
    |> Enum.map(fn {x, y, _} -> {x, y} end)
  end

  defp transpose(board, player) do
    # Transpose the board if this is player X so that we don't
    # have to care later about which side of the board to start
    # on.
    cond do
      player == "O" -> board
      true ->
        board
        |> Enum.map(&String.codepoints/1)
        |> List.zip
        |> Enum.map(&Enum.join(Tuple.to_list(&1)))
    end
  end

  defp build_board(board) do
    d = Enum.count(board) - 1
    for x <- 0..d,
        y <- 0..d,
        v = board |> Enum.at(x) |> String.at(y),
        do: {x, y, v}
  end

  # You can't win if you don't play.
  defp winner?([], _), do: false
  defp winner?(pieces, goal_x) do
    pieces
    # Find the starting pieces.
    |> Enum.filter(fn {x, _} -> x == 0 end)
    # Start following the adjacent pieces.
    |> Enum.any?(&(adjacent?(&1, pieces, goal_x)))
  end

  defp adjacent?(cell={x, _}, pieces, goal_x) do
    next = get_adjacents(cell, pieces)
    cond do
      # We managed to reach the goal.
      goal_x == x           -> true
      # Nothing more in this chain.
      Enum.count(next) == 0 -> false
      # Continue down the chain, remove ourselves so we can't get a
      # circular reference.
      true                  ->
        Enum.any?(next, &(adjacent?(&1, List.delete(pieces, cell), goal_x)))
    end
  end

  defp get_adjacents({x, y}, pieces) do
    # Hooray for hexagons.
    adjacent_coords = [
      {x + 1, y},
      {x - 1, y},
      {x, y + 1},
      {x, y - 1},
      {x + 1, y - 1},
      {x - 1, y + 1}
    ]
    Enum.filter(pieces, &(&1 in adjacent_coords))
  end
end

File: ruby/connect/connect.rb
class Board
  def initialize(board)
    board_o = board.map { |row| row.gsub(/ /, '').split('') }
    board_x = transpose_board(board_o, 'X')
    @pieces_o = pieces_for(build_board(board_o), 'O')
    @pieces_x = pieces_for(build_board(board_x), 'X')
    @goal_x = board_o.first.length - 1
  end

  def winner
    return 'X' if winner?(@pieces_x)
    return 'O' if winner?(@pieces_o)
    ""
  end

  def winner?(pieces)
    starting_pieces = pieces.select { |p| p.first == 0 }
    starting_pieces.any? { |p| adjacent?(p, pieces) }
  end

  def adjacent?(cell, pieces)
    (x, y, v) = cell
    return true if x == @goal_x
    next_pieces = get_adjacents(x, y, pieces)
    return false if next_pieces.length == 0
    pieces.delete(cell)
    next_pieces.any? { |np| adjacent?(np, pieces) }
  end

  def get_adjacents(x, y, pieces)
    adjacent_coords = [
      [x + 1, y],
      [x - 1, y],
      [x, y + 1],
      [x, y - 1],
      [x + 1, y - 1],
      [x - 1, y + 1]
    ]
    pieces.select do |(px, py, pv)|
      adjacent_coords.any? { |(ax, ay)| px == ax && py == ay  }
    end
  end

  def pieces_for(board, player)
    board.select { |p| p.last == player }
  end

  def build_board(board)
    pieces = []
    d = board.length - 1
    (0..d).each do |x|
      (0..d).each do |y|
        v = board[x][y]
        pieces.push([x, y, v])
      end
    end
    pieces
  end

  def transpose_board(board, player)
    board.transpose if player == 'X'
  end
end

module BookKeeping
  VERSION = 1
end
